include/aws/cal/der.h

```diff
@@ -0,0 +1,249 @@
+
+/**
+ * Begins a SET of objects in the DER stream
+ * @param encoder The encoder to use
+ * @return AWS_OP_ERR if an error occurs, otherwise AWS_OP_SUCCESS
+ */
+AWS_CAL_API int aws_der_encoder_begin_set(struct aws_der_encoder *encoder);
+
+/**
+ * Finishes a SET and applies it to the DER stream buffer
+ * @param encoder The encoder to update
+ * @return AWS_OP_ERR if an error occurs, otherwise AWS_OP_SUCCESS
+ */
+AWS_CAL_API int aws_der_encoder_end_set(struct aws_der_encoder *encoder);
+
+/**
+ * Retrieves the contents of the encoder stream buffer
+ * @param encoder The encoder to read from
+ * @param cursor The cursor to point at the stream buffer
+ * @return AWS_OP_ERR if an error occurs, otherwise AWS_OP_SUCCESS
+ */
+AWS_CAL_API int aws_der_encoder_get_contents(struct aws_der_encoder *encoder, struct aws_byte_cursor *contents);
+
+/**
+ * Initializes an DER decoder
+ * @param decoder The decoder to initialize
+ * @param allocator The allocator to use
+ * @param buffer The DER formatted buffer to parse
+ * @return AWS_OP_ERR if an error occurs, otherwise AWS_OP_SUCCESS
+ */
+AWS_CAL_API int aws_der_decoder_init(
+    struct aws_der_decoder *decoder,
+    struct aws_allocator *allocator,
+    struct aws_byte_buf *buffer);
+
+/**
+ * Cleans up a DER encoder
+ * @param decoder The encoder to clean up
+ */
+AWS_CAL_API void aws_der_decoder_clean_up(struct aws_der_decoder *decoder);
+
+/**
+ * Parses the internal buffer into a TLV or sequence/set of TLVs
+ * @param decoder The decoder to parse
+ * @return AWS_OP_ERR if an error occurs, otherwise AWS_OP_SUCCESS
+ */
+AWS_CAL_API int aws_der_decoder_parse(struct aws_der_decoder *decoder);
+
+/**
+ * Allows for iteration over the decoded TLVs.
+ * @param decoder The decoder to iterate over
+ * @return true if there is a tlv to read after advancing, false when done
+ */
+AWS_CAL_API bool aws_der_decoder_next(struct aws_der_decoder *decoder);
+
+/**
+ * The type of the current TLV
+ * @param decoder The decoder to inspect
+ * @return AWS_OP_ERR if an error occurs, otherwise AWS_OP_SUCCESS
+ */
+AWS_CAL_API enum aws_der_type aws_der_decoder_tlv_type(struct aws_der_decoder *decoder);
+
+/**
+ * The size of the current TLV
+ * @param decoder The decoder to inspect
+ * @return AWS_OP_ERR if an error occurs, otherwise AWS_OP_SUCCESS
+ */
+AWS_CAL_API size_t aws_der_decoder_tlv_length(struct aws_der_decoder *decoder);
+
+/**
+ * The number of elements in the current TLV container
+ * @param decoder The decoder to inspect
+ * @return Number of elements in the current container
+ */
+AWS_CAL_API size_t aws_der_decoder_tlv_count(struct aws_der_decoder *decoder);
+
+/**
+ * Extracts the current TLV string value (BIT_STRING, OCTET_STRING)
+ * @param decoder The decoder to extract from
+ * @param string The buffer to store the string into
+ */
+AWS_CAL_API void aws_der_decoder_tlv_string(struct aws_der_decoder *decoder, struct aws_byte_cursor *string);
```

Comment:
For all these getters, you could just return the value:
`struct aws_byte_cursor aws_der_decoder_tlv_string(struct aws_der_decoder *decoder);`

if you had an error code, there would be a good reason to use an out-param, but this is a void fn

Reply:
So, I ended up implementing these as asserts, but they should be error code returning functions. I will fix that.